File: ruby/text-to-slack-emoji.rb
# When passed (1) a style name and (2) a string of supported characters, this script
# converts the string to Slack emoji symbols of the requested style. Case is ignored.
# The requested character emojis must, of course, be available in Slack.
# Update SUPPORTED_CHARSETS as needed to match the emojis available in Slack.
#
# Run: ruby slack-text-converter.rb STYLE_NAME "STRING_OF_SUPPORTED_CHARS"
#        (Example: ruby slack-text-converter.rb cookie "Sprint planning")
#
# Tip: On macOS, append ` | pbcopy` to output directly to your clipboard!

LETTERS = *('a'..'z')
NUMBERS = *('0'..'9')
SPACE = ' '

SUPPORTED_CHARSETS = {
  cookie:
    {
      chars: [*LETTERS, '!', '?', '&', '•', SPACE],
      converter: ->(char) do
        case char
        when SPACE then ':blank:'
        when '!' then ':cookie-exclaim:'
        when '?' then ':cookie-question:'
        when '&' then ':cookie-and:'
        when '•' then ':cookie-dot:'
        else ":cookie-#{char}:"
        end
      end
    },
  bluebox:
    {
      chars: [*LETTERS, SPACE],
      converter: ->(char) do
        case char
        when SPACE then ':blank: '
        when 'o' then ':alpha-0: '
        else ":alpha-#{char}: "
        end
      end
    },
  alphawhite:
    {
      chars: [*LETTERS, '!', '?', '#', '@', SPACE],
      converter: ->(char) do
        case char
        when SPACE then ':blank:'
        when '!' then ':alphabet-white-exclamation:'
        when '?' then ':alphabet-white-question:'
        when '#' then ':alphabet-white-hash:'
        when '@' then ':alphabet-white-at:'
        else ":alphabet-white-#{char}:"
        end
      end
    },
  alphayellow:
    {
      chars: [*LETTERS, '!', '?', '#', '@', 'ñ', SPACE],
      converter: ->(char) do
        case char
        when SPACE then ':blank:'
        when '!' then ':alphabet-yellow-exclamation:'
        when '?' then ':alphabet-yellow-question:'
        when '#' then ':alphabet-yellow-hash:'
        when '@' then ':alphabet-yellow-at:'
        when 'ñ' then ':alphabet-yellow-nñ:'
        else ":alphabet-yellow-#{char}:"
        end
      end
    },
  alphasnow:
    {
      chars: [*LETTERS, SPACE],
      converter: ->(char) do
        case char
        when SPACE then ':blank:'
        when 'i' then ':alpha_snow_i2:'
        else ":alpha_snow_#{char}:"
        end
      end
    },
  tiles:
    {
      chars: [*LETTERS, 'ñ', SPACE],
      converter: ->(char) do
        case char
        when SPACE then ':letter_blank:'
        when 'ñ' then ':letterñ:'
        else ":letter_#{char}:"
        end
      end
    },
  pokemon:
    {
      chars: [*LETTERS, SPACE],
      converter: ->(char) do
        case char
        when SPACE then ':blank:'
        else ":pokemonfont-#{char}:"
        end
      end
    },
  merah:
    {
      chars: [*LETTERS, SPACE],
      converter: ->(char) do
        case char
        when SPACE then ':blank:'
        # The "merah" set is incomplete and inconsistently named.
        when 'b' then ':merahbb:'
        when 'd' then ':merahdd:'
        when 'e' then ':merahee:'
        when 'f' then ':ff:'
        when 'h' then ':merahhh:'
        when 'k' then ':merahkk:'
        when 'l' then ':merahll:'
        when 'q' then ':alpha-q:'
        when 'r' then ':merahrr:'
        when 's' then ':merahsss:'
        when 'u' then ':merahuuu:'
        when 'x' then ':alpha-x:'
        when 'y' then ':merahhyyy:'
        when 'z' then ':alpha-z:'
        else ":merah#{char}:"
        end
      end
    },
  magazine:
    {
      chars: [*LETTERS, SPACE],
      converter: ->(char) do
        case char
        when SPACE then ':blank:'
        else ":magazine_#{char}:"
        end
      end
    },
  custom:
    {
      chars: [*LETTERS, *NUMBERS, '?', '!', SPACE],
      converter: ->(char) do
        case char
        when SPACE then ':blank:'
        when 'a' then ":m-a:"
        when 'b' then ":b4:"
        when 'c' then ":c2:"
        when 'd' then ":devo-d:"
        when 'e' then ":edge:"
        when 'f' then ":ff:"
        when 'g' then ":g-neon:"
        when 'h' then ":h:"
        when 'i' then ":info:"
        when 'j' then ":super-j:"
        when 'k' then ":m'kay:"
        when 'l' then ":labsslide-1:"
        when 'm' then ":m:"
        when 'n' then ":n64:"
        when 'o' then ":o:"
        when 'p' then ":p2:"
        when 'q' then ":qflash:"
        when 'r' then ":r:"
        when 's' then ":scon:"
        when 't' then ":kid-t:"
        when 'u' then ":m-u:"
        when 'v' then ":devo-v:"
        when 'w' then ":walphabet:"
        when 'x' then ":x:"
        when 'y' then ":y1:"
        when 'z' then ":zelle_onfire:"
        when '0' then ":0_bats:"
        when '1' then ":number-1-red:"
        when '2' then ":number2:"
        when '3' then ":number-3-flip:"
        when '4' then ":mana-4:"
        when '5' then ":round-red-5:"
        when '6' then ":mana-6:"
        when '7' then ":7-up:"
        when '8' then ":8flower:"
        when '9' then ":9lego:"
        when '!' then ":exclamation:"
        when '?' then ":question-icon:"
        else char # Should not be reached.
        end
      end
    },
  numbers:
    {
      chars: [*NUMBERS, SPACE],
      converter: ->(char) do
        case char
        when SPACE then ':blank:'
        else ":num#{char}:"
        end
      end
    },
  squarenumbers:
    {
      chars: [*NUMBERS, SPACE],
      converter: ->(char) do
        case char
        when SPACE then ':blank:'
        when '0' then ':zero:'
        when '1' then ':one:'
        when '2' then ':two:'
        when '3' then ':three:'
        when '4' then ':four:'
        when '5' then ':five:'
        when '6' then ':six:'
        when '7' then ':seven:'
        when '8' then ':eight:'
        else ':nine:'
        end
      end
    }
  }

supported_styles = SUPPORTED_CHARSETS.keys

if ARGV.count.zero? || ARGV.count > 2
  STDERR.puts "Pass in (1) a style name and (2) a string containing only supported characters for that style."
  STDERR.puts "Supported styles: #{supported_styles.join(', ')}"
  return
end

style = ARGV[0].downcase.to_sym

def style_chars(style, separator)
  SUPPORTED_CHARSETS[style][:chars]
    .map { |c| c == SPACE ? '(space)' : c }
    .join(separator)
end

is_supported_style = supported_styles.include? style

if ARGV.count == 1 && is_supported_style
  STDERR.puts "The \"#{style}\" style supports the following characters:"
  STDERR.puts style_chars(style, SPACE)
  return
end

unless is_supported_style
  STDERR.puts "\"#{style}\" is not a supported style."
  STDERR.puts "Supported styles: #{supported_styles.join(', ')}"
  return
end

lowered_chars = ARGV[1].downcase.chars

if lowered_chars.length.zero?
  STDERR.puts "You must enter text for conversion."
  return
end

is_supported_char = ->(c) { SUPPORTED_CHARSETS[style][:chars].include? c }

unless lowered_chars.all?(&is_supported_char)
  STDERR.puts "Only the following characters are supported for the \"#{style}\" style:"
  STDERR.puts style_chars(style, SPACE)
  return
end

style_converter = SUPPORTED_CHARSETS[style][:converter]
puts lowered_chars.map(&style_converter).join

File: ocaml/text_to_slack_emoji.ml
let (<|) f x = f x

let space = ' '

module Styles = struct
    let letters = List.init 26 (fun i -> Char.chr (Char.code 'a' + i)) (* [ 'a' .. 'z' ] *)
    let numbers = List.init 10 (fun i -> Char.chr (Char.code '0' + i))

    type style = {
        name: string;
        supported_chars: char list;
        converter: char -> string
    }

    let styles =
        [
            { name = "cookie";
              supported_chars = letters @ [space; '!'; '?'; '&'; (* '•' *)];
              converter =
                    fun ch ->
                        match ch with
                        | c when c = space -> ":blank:"
                        | '!' -> ":cookie-exclaim:"
                        | '?' -> ":cookie-question:"
                        | '&' -> ":cookie-and:"
                        (* | '•' -> ":cookie-dot:" *)
                        | c -> Printf.sprintf ":cookie-{%c}:" c
            };
            { name = "bluebox";
              supported_chars = letters @ [space];
              converter =
                    fun ch ->
                        match ch with
                        | c when c = space -> ":blank: "
                        | 'o' -> ":alpha-0: "
                        | c -> Printf.sprintf ":alpha-{%c}:" c (* Hairspace *)
            };
            { name = "alphawhite";
              supported_chars = letters @ [space; '!'; '?'; '#'; '@'];
              converter =
                    fun ch ->
                        match ch with
                        | c when c = space -> ":blank:"
                        | '!' -> ":alphabet-white-exclamation:"
                        | '?' -> ":alphabet-white-question:"
                        | '#' -> ":alphabet-white-hash:"
                        | '@' -> ":alphabet-white-at:"
                        | c -> Printf.sprintf ":alphabet-white-{%c}:" c
            };
            { name = "alphayellow";
              supported_chars = letters @ [space; '!'; '?'; '#'; '@'; (* 'ñ' *)];
              converter =
                    fun ch ->
                        match ch with
                        | c when c = space -> ":blank:"
                        | '!' -> ":alphabet-yellow-exclamation:"
                        | '?' -> ":alphabet-yellow-question:"
                        | '#' -> ":alphabet-yellow-hash:"
                        | '@' -> ":alphabet-yellow-at:"
                        (* | 'ñ' -> ":alphabet-yellow-nñ:" *)
                        | c -> Printf.sprintf ":alphabet-yellow-{%c}:" c
            };
            { name = "alphasnow";
              supported_chars = letters @ [space];
              converter =
                    fun ch ->
                        match ch with
                        | c when c = space -> ":blank:"
                        | 'i' -> ":alpha_snow_i2:"
                        | c -> Printf.sprintf ":alpha_snow_{%c}:" c
            };
            { name = "tiles";
              supported_chars = letters @ [space; (* 'ñ' *)];
              converter =
                    fun ch ->
                        match ch with
                        | c when c = space -> ":letter_blank:"
                        (* | 'ñ' -> ":letterñ:" *)
                        | c -> Printf.sprintf ":letter_{%c}:" c
            };
            { name = "pokemon";
              supported_chars = letters @ [space];
              converter =
                    fun ch ->
                        match ch with
                        | c when c = space -> ":blank:"
                        | c -> Printf.sprintf ":pokemonfont-{%c}:" c
            };
            { name = "merah";
              supported_chars = letters @ [space];
              converter =
                    fun ch ->
                        match ch with
                        | c when c = space -> ":blank:"
                        (* The "merah" set is incomplete and inconsistently named. *)
                        | 'b' -> ":merahbb:"
                        | 'd' -> ":merahdd:"
                        | 'e' -> ":merahee:"
                        | 'f' -> ":ff:"
                        | 'h' -> ":merahhh:"
                        | 'k' -> ":merahkk:"
                        | 'l' -> ":merahll:"
                        | 'q' -> ":alpha-q:"
                        | 'r' -> ":merahrr:"
                        | 's' -> ":merahsss:"
                        | 'u' -> ":merahuuu:"
                        | 'x' -> ":alpha-x:"
                        | 'y' -> ":merahhyyy:"
                        | 'z' -> ":alpha-z:"
                        | c -> Printf.sprintf ":merah{%c}:" c
            };
            { name = "magazine";
              supported_chars = letters @ [space];
              converter =
                    fun ch ->
                        match ch with
                        | c when c = space -> ":blank:"
                        | c -> Printf.sprintf ":magazine_{%c}:" c
            };
            { name = "custom";
              supported_chars = letters @ numbers @ [space; '?'; '!'];
              converter =
                    fun ch ->
                        match ch with
                        | c when c = space -> ":blank:"
                        | 'a' -> ":m-a:"
                        | 'b' -> ":b4:"
                        | 'c' -> ":c2:"
                        | 'd' -> ":devo-d:"
                        | 'e' -> ":edge:"
                        | 'f' -> ":ff:"
                        | 'g' -> ":g-neon:"
                        | 'h' -> ":h:"
                        | 'i' -> ":info:"
                        | 'j' -> ":super-j:"
                        | 'k' -> ":m'kay:"
                        | 'l' -> ":labsslide-1:"
                        | 'm' -> ":m:"
                        | 'n' -> ":n64:"
                        | 'o' -> ":o:"
                        | 'p' -> ":p2:"
                        | 'q' -> ":qflash:"
                        | 'r' -> ":r:"
                        | 's' -> ":scon:"
                        | 't' -> ":kid-t:"
                        | 'u' -> ":m-u:"
                        | 'v' -> ":devo-v:"
                        | 'w' -> ":walphabet:"
                        | 'x' -> ":x:"
                        | 'y' -> ":y1:"
                        | 'z' -> ":zelle_onfire:"
                        | '0' -> ":0_bats:"
                        | '1' -> ":number-1-red:"
                        | '2' -> ":number2:"
                        | '3' -> ":number-3-flip:"
                        | '4' -> ":mana-4:"
                        | '5' -> ":round-red-5:"
                        | '6' -> ":mana-6:"
                        | '7' -> ":7-up:"
                        | '8' -> ":8flower:"
                        | '9' -> ":9lego:"
                        | '!' -> ":exclamation:"
                        | '?' -> ":question-icon:"
                        | _ -> String.make 1 ch (* Should not be reached. *)
            };
            { name = "numbers";
              supported_chars = numbers @ [space];
              converter =
                    fun ch ->
                        match ch with
                        | c when c = space -> ":blank:"
                        | c -> Printf.sprintf ":num{%c}:" c
            };
            { name = "squarenumbers";
              supported_chars = numbers @ [space];
              converter =
                    fun ch ->
                        match ch with
                        | c when c = space -> ":blank:"
                        | '0' -> ":zero:"
                        | '1' -> ":one:"
                        | '2' -> ":two:"
                        | '3' -> ":three:"
                        | '4' -> ":four:"
                        | '5' -> ":five:"
                        | '6' -> ":six:"
                        | '7' -> ":seven:"
                        | '8' -> ":eight:"
                        | _ -> "nine"
            };
        ]

    let styleNames = styles |> List.map (fun s -> s.name)

end

module ArgValidation = struct
    (* open Result.Let_syntax *)
    open Styles

    type user_args = { style: string; text: string }

    let validate =
        let supportedStyleNames =
            String.concat " " (styles |> List.map (fun s -> s.name)) in


        let rawArgs =
            Sys.argv
            |> Array.to_list
            |> List.tl (* The head contains the script filename. *) in

        let argCount (rawArgs:string list) =
            let errorText =
                String.concat
                    "\n"
                    ["Pass in (1) a style name and (2) a string containing only supported characters for that style.";
                     Printf.sprintf "Supported styles: %s" supportedStyleNames] in

            match List.length rawArgs with
            | l when l = 2 ->
                Ok {
                    style = rawArgs |> List.hd |> String.lowercase_ascii;
                    text = List.nth rawArgs 1 |> String.lowercase_ascii
                }
            | _ -> Error errorText in

        let styleName args =
            let errorText =
                String.concat
                    "\n"
                    [Printf.sprintf "Style \"%s\" not found." args.style;
                     Printf.sprintf "Supported styles: %s" supportedStyleNames] in

            match styleNames |> List.mem args.style with
            | true -> Ok args
            | false -> Error errorText in

        let inputLength args =
            match String.length args.text with
            | 0 -> Error "You must enter text to be converted."
            | _ -> Ok args in

        let inputChars args =
            let style =
                styles
                |> List.filter (fun s -> s.name = args.style)
                |> List.hd in

            let ensure_visible_char ch =
                if ch = space then "<SPACE>" else String.make 1 ch in

            let invalidChars =
                args.text
                |> String.to_seq
                |> List.of_seq
                |> List.filter (fun ch -> not <| List.mem ch style.supported_chars)
                |> List.map ensure_visible_char in

            let error style =
              let supported_chars style =
                let chars = style.supported_chars |> List.map ensure_visible_char in
                String.concat " " chars in

              String.concat
                "\n"
                [Printf.sprintf "Invalid characters found for style \"%s\": %s" (style.name) (String.concat " " invalidChars);
                 Printf.sprintf "Supported characters: %s" (supported_chars style)] in

            if List.length invalidChars = 0
            then Ok args
            else Error (error style) in

        let ( let* ) = Result.bind in
        let* args = argCount rawArgs in
        let* args' = styleName args in
        let* args'' = inputLength args' in
        let* args''' = inputChars args'' in
        Ok args'''
end

open ArgValidation
open Styles

let args = validate

let get_style args =
    styles
    |> List.filter (fun s -> s.name = args.style)
    |> List.hd

let convert_text args style =
  let text : string = args.text in
  let converter : char -> string = style.converter in
  text
  |> String.to_seq
  |> Seq.map converter
  |> List.of_seq
  |> fun lst -> String.concat "" lst

let () =
  match args with
  | Error e ->
      Printf.eprintf "%s\n" e
  | Ok a ->
      a
      |> get_style
      |> (convert_text a)
      |> fun s -> Printf.printf "%s\n" s
